File: frontend/src/lib/api/traces.ts
/**
 * Typed client for POST /v1/traces.
 *
 * Mirrors backend/src/app/schemas/traces.py. The backend forbids unknown
 * fields on a trace entry rather than ignoring them, so this type drifting
 * from the Python schema produces a loud 422 on the next upload instead of an
 * archive that silently stopped recording a field.
 *
 * Called from the background service worker, like classify and explain, so
 * the API base URL stays in one place.
 */

/** Matches MAX_TRACE_ENTRIES in the backend schema. Trimmed here so an
 * unusually large page produces a truncated capture rather than a rejected
 * one -- a partial trace is worth more than none. */
export const MAX_TRACE_ENTRIES = 5000;

export interface TraceEntryPayload {
  id: string;
  text: string;
  tag: string;
  role: string;
  step: string | null;
  selector: string;
  ruleHits: string[];
  sentToModel: boolean;
  /** null = never resolved, [] = confirmed benign, non-empty = labels found. */
  findingLabels: string[] | null;
  firstSeenAt: number;
  lastSeenAt: number;
}

export interface StoreTraceRequest {
  scan_id: string;
  url: string;
  page_score: number;
  entries: TraceEntryPayload[];
}

export interface StoreTraceResponse {
  scan_id: string;
  object_key: string;
  bucket: string;
  replaced: boolean;
  entry_count: number;
}

export class TraceApiError extends Error {
  constructor(
    message: string,
    public status: number,
  ) {
    super(message);
    this.name = "TraceApiError";
  }
}

export interface TraceClientOptions {
  baseUrl: string;
  fetchImpl?: typeof fetch;
}

export function createTraceClient({ baseUrl, fetchImpl = fetch }: TraceClientOptions) {
  return {
    async store(req: StoreTraceRequest): Promise<StoreTraceResponse> {
      const res = await fetchImpl(`${baseUrl}/v1/traces`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          ...req,
          entries: req.entries.slice(0, MAX_TRACE_ENTRIES),
        }),
      });

      if (!res.ok) {
        let detail = "";
        try {
          const body = (await res.json()) as { detail?: string };
          detail = typeof body.detail === "string" ? body.detail : "";
        } catch {
          detail = await res.text().catch(() => "");
        }
        throw new TraceApiError(
          detail || `POST /v1/traces failed: ${res.status}`,
          res.status,
        );
      }

      return res.json() as Promise<StoreTraceResponse>;
    },
  };
}

File: backend/src/app/schemas/traces.py
"""Request/response contracts for /v1/traces.

A trace is what the extension already assembles for its own debugging (see
`exportTrace` in the content script): every candidate the page produced, with
its tag, role, selector, whether it was sent to the model, and what came back.
Storing it is what turns "I noticed a gap on some page once" into something
that can be re-examined without finding and re-scanning the page.

The entry shape here is deliberately permissive about extra keys being absent,
but not about unknown ones: the extension and this schema must not drift
silently, since a field quietly dropped on upload is only discovered much later
when the archive turns out not to contain it.
"""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

#: Bounded so one page cannot post an unbounded body. Real pages top out
#: around 600 candidates (MAX_SNIPPETS_PER_PAGE in the extension); the headroom
#: covers extraction passes that saw more before deduping.
MAX_TRACE_ENTRIES = 5000


class TraceEntry(BaseModel):
    """One candidate's full journey, as the content script recorded it."""

    model_config = ConfigDict(extra="forbid")

    id: str
    text: str
    tag: str = "span"
    role: str = "none"
    step: str | None = None
    selector: str = ""
    ruleHits: list[str] = Field(default_factory=list)
    sentToModel: bool = False
    #: null = never resolved, [] = confirmed benign, non-empty = labels found.
    #: The three are genuinely different and collapsing them would destroy the
    #: main reason to keep the trace at all.
    findingLabels: list[str] | None = None
    firstSeenAt: int = 0
    lastSeenAt: int = 0


class StoreTraceRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    #: Stable per page load. Re-uploading the same scan after more batches
    #: resolve replaces the earlier, less complete capture rather than adding
    #: a near-duplicate.
    scan_id: Annotated[str, Field(min_length=1, max_length=100)]

    url: Annotated[str, Field(min_length=1, max_length=2000)]

    page_score: Annotated[int, Field(ge=0, le=100)] = 0

    entries: Annotated[list[TraceEntry], Field(max_length=MAX_TRACE_ENTRIES)]

    @field_validator("url")
    @classmethod
    def _http_only(cls, v: str) -> str:
        # Keeps chrome://, file:// and data: captures out of the archive.
        # Nothing useful is learned from them and file:// paths in particular
        # leak local directory structure into stored records.
        if not v.startswith(("http://", "https://")):
            raise ValueError("url must be http(s)")
        return v


class StoreTraceResponse(BaseModel):
    scan_id: str
    object_key: str
    bucket: str
    #: True when this replaced an earlier capture of the same scan.
    replaced: bool = False
    entry_count: int


class TraceSummary(BaseModel):
    """One row of the index. Deliberately not the trace itself -- listing
    should stay cheap regardless of how large the stored objects are."""

    scan_id: str
    object_key: str
    host: str
    url: str
    captured_at: str
    candidate_count: int
    flagged_count: int
    page_score: int
    labels: list[str]


class TraceListResponse(BaseModel):
    traces: list[TraceSummary]
    total_indexed: int


TraceSortOrder = Literal["newest"]

__all__ = [
    "MAX_TRACE_ENTRIES",
    "StoreTraceRequest",
    "StoreTraceResponse",
    "TraceEntry",
    "TraceListResponse",
    "TraceSummary",
]
